packages/protocol/contracts/stability/Exchange.sol

```diff
@@ -113,11 +134,68 @@ contract Exchange is
     updateBucketsIfNecessary
     nonReentrant
     returns (uint256)
+  {
+    return _sell(sellAmount, minBuyAmount, sellGold);
+  }
+
+  /**
+   * @notice Exchanges sellAmount of sellToken in exchange for at least minBuyAmount of buyToken
+   * Requires the sellAmount to have been approved to the exchange
+   * @param sellAmount The amount of sellToken the user is selling to the exchange
+   * @param minBuyAmount The minimum amount of buyToken the user has to receive for this
+   * transaction to succeed
+   * @param sellGold `true` if gold is the sell token
+   * @return The amount of buyToken that was transfered
+   */
+  function _sell(uint256 sellAmount, uint256 minBuyAmount, bool sellGold)
+    private
+    returns (uint256)
   {
     uint256 buyAmount = _getBuyTokenAmount(sellAmount, sellGold);
 
     require(buyAmount >= minBuyAmount, "Calculated buyAmount was less than specified minBuyAmount");
 
+    _exchangeAmounts(sellAmount, buyAmount, sellGold);
+    return buyAmount;
+  }
+
+  /**
+   * @notice Exchanges buyAmount of buyToken in exchange for at most maxSellAmount of sellToken
+   * Requires the sellAmount to have been approved to the exchange
+   * @param buyAmount The amount of buyToken the user is buying from the exchange
+   * @param maxSellAmount The maximum amount of sellToken the user can sell for this
+   * transaction to succeed
+   * @param buyGold `true` if gold is the buy token
+   * @return The amount of sellToken that was sold
+   * @dev This function can be frozen using the Freezable interface.
+   */
+  function buy(uint256 buyAmount, uint256 maxSellAmount, bool buyGold)
+    external
+    onlyWhenNotFrozen
+    updateBucketsIfNecessary
+    nonReentrant
+    returns (uint256)
+  {
+    bool sellGold = !buyGold;
+    uint256 sellAmount = _getSellTokenAmount(buyAmount, sellGold);
+
+    require(
+      sellAmount <= maxSellAmount,
+      "Calculated sellAmount was greater than specified maxSellAmount"
+    );
+
+    _exchangeAmounts(sellAmount, buyAmount, sellGold);
+    return sellAmount;
+  }
+
+  /**
+   * @notice Exchanges sellAmount of sellToken in exchange for buyAmount of buyToken
```

Comment:
```
@notice Exchanges a specific amount of one token for a specific amount of another.
@param sellAmount The number of tokens to send to the exchange.
@param buyAmount The number of tokens for the exchange to send in return.
@param sellGold True if the msg.sender is sending CELO to the exchange, false otherwise.
```